### PR TITLE
Nonlinear training speedup & group features

### PR DESCRIPTION
**Pull request by rohskopf**
This PR makes huge improvements in training performance. We introduce a new `Configuration` class that contains _assignments_ (not copies, to save RAM) to data in the shared arrays and fitsnap dicts. This drastically reduces the overhead in the dataloader as mentioned in https://github.com/FitSNAP/FitSNAP/pull/119, since we only need to collate a list of `Configuration` objects instead of many large arrays with complicated indexing. This results in near O(N) training performance, where N is number of atoms in the training set; strictly speaking, scaling may be N times number of neighbors. The speedup is therefore more notable as data set size is increased, e.g. the screenshots below show a 50x speedup when training to 10k configs in the entropy set (~50 atoms per config), with the same settings (batch size = 4, etc) used for each fit:

image

Additionally, this PR adds some desirable features that improve nonlinear training:

1. Parallelized configuration processing for nonlinear solvers.
2. Group weights that are overridden by global weights for all configs if desired.
3. Group training/testing fractions that are overridden by global fractions for all configs if desired.
4. Removed dependence on LAMMPS ML-IAP libraries for writing LAMMPS-ready PyTorch files for performing MD; this allows everyone to create LAMMPS-ready models regardless of their Python setup.

These new features are explained in the PyTorch docs: https://fitsnap.github.io/Pytorch.html

Avenues for further improved time & space performance:

- Create shared arrays that are PyTorch tensors instead of numpy arrays; this will reduce overhead associated with converting to PyTorch tensors every epoch. 
- Multi node training will increase the RAM capacity.
- Not storing descriptor gradients for configs with a force weight of zero will increase RAM capacity. 
- Optionally storing a pickled list of Configuration objects for loading previously computed descriptors and descriptor gradients.

### CODE REVIEW

**Review by mitwood — APPROVED** [review]
The advance here is undeniable, who doesn't want a faster code?
